stdlib/path.gr

```diff
@@ -0,0 +1,729 @@
+  match (path) {
+    (_, _, [name, ..._]) => Some(name),
+    _ => None,
+  }
+
+/**
+ * Retrieves the basename (named final segment) of a path.
+ * 
+ * @param path: The path to inspect
+ * @returns `Some(path)` containing the basename of the path, or `None` if the path does not have one
+ * 
+ * @example basename(fromString("./dir/file.txt")) == Some("file.txt")
+ * @example basename(fromString(".."))) == None
+ * 
+ * @since v0.5.5
+ */
+export let basename = (path: Path) => {
+  basename(pathInfo(path))
+}
+
+// should only be used on file paths
+let stem = (path: PathInfo) =>
+  match (path) {
+    (_, _, [name, ..._]) => {
+      match (String.lastIndexOf(".", name)) {
+        Some(dotI) => {
+          let len = String.length(name)
+          if (dotI > 0 && dotI < len - 1) String.slice(0, dotI, name) else name
+        },
+        None => name,
+      }
+    },
+    _ => "",
+  }
+
+/**
+ * Retrieves the basename of a file path without the extension.
+ * 
+ * @param path: The path to inspect
+ * @returns `Ok(path)` containing the stem of the file path, or `Err(err)` if the path is a directory path
+ * 
+ * @example stem(fromString("file.txt")) == Ok("file")
+ * @example stem(fromString(".gitignore")) == Ok(".gitignore")
+ * @example stem(fromString("/dir/")) == Err(IncompatiblePathType) // can only take stem of a file path
+ * 
+ * @since v0.5.5
+ */
+export let stem = (path: Path) => {
+  match (pathInfo(path)) {
+    (_, Dir, _) => Err(IncompatiblePathType),
+    pathInfo => Ok(stem(pathInfo)),
+  }
+}
+
+// should only be used on file paths
+let extension = (path: PathInfo) =>
+  match (path) {
+    (_, _, [name, ..._]) => {
+      match (String.lastIndexOf(".", name)) {
+        Some(dotI) => {
+          let len = String.length(name)
+          if (dotI > 0 && dotI < len - 1) String.slice(dotI, len, name) else ""
+        },
+        None => "",
+      }
+    },
+    _ => "",
+  }
+
+/**
+ * Retrieves the extension on the basename of a file path.
+ * 
+ * @param path: The path to inspect
+ * @returns `Ok(path)` containing the extension of the file path, or `Err(err)` if the path is a directory path
+ * 
+ * @example extension(fromString("file.txt")) == Ok(".txt")
+ * @example extension(fromString(".gitignore")) == Ok("")
+ * @example extension(fromString("/dir/")) == Err(IncompatiblePathType) // can only take extension of a file path
+ * 
+ * @since v0.5.5
+ */
+export let extension = (path: Path) => {
+  match (pathInfo(path)) {
+    (_, Dir, _) => Err(IncompatiblePathType),
+    pathInfo => Ok(extension(pathInfo)),
+  }
+}
+
+// should only be used on absolute paths
+let root = (path: PathInfo) =>
+  match (path) {
+    (Abs(root), _, _) => root,
+    _ => fail "Impossible: malformed absolute path data",
+  }
+
+/**
+ * Retrieves the root of the absolute path
+ * 
+ * @param path: The path to inspect
+ * @returns `Ok(root)` containing the root of the path, or `Err(err)` if the path is a relative path
```

Comment:
```suggestion
 * @returns `Ok(root)` containing the root of the path or `Err(err)` if the path is a relative path
```

Reply:
I don't know what happened to my grammar skills xD